sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PushdownPredicatesAndPruneColumnsForCTEDef.scala

```diff
@@ -0,0 +1,171 @@
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import scala.collection.mutable
+
+import org.apache.spark.sql.catalyst.expressions.{And, Attribute, AttributeSet, Expression, Literal, Or, SubqueryExpression}
+import org.apache.spark.sql.catalyst.planning.ScanOperation
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern.CTE
+
+/**
+ * Infer predicates and column pruning for [[CTERelationDef]] from its reference points, and push
+ * the disjunctive predicates as well as the union of attributes down the CTE plan.
+ */
+object PushdownPredicatesAndPruneColumnsForCTEDef extends Rule[LogicalPlan] {
+
+  // CTE_id - (CTE_definition, precedence, predicates_to_push_down, attributes_to_prune)
+  private type CteMap = mutable.HashMap[Long, (CTERelationDef, Int, Seq[Expression], AttributeSet)]
+
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    if (!plan.isInstanceOf[Subquery] && plan.containsPattern(CTE)) {
+      val cteMap = new CteMap
+      gatherPredicatesAndAttributes(plan, cteMap)
+      pushdownPredicatesAndAttributes(plan, cteMap)
+    } else {
+      plan
+    }
+  }
+
+  private def reverseMapExpressions(
+      input: Seq[Expression],
+      mapping: Map[Attribute, Expression]): Seq[Expression] = {
+    input.map(e => e.transform {
+      case a: Attribute =>
+        mapping.keys.find(_.semanticEquals(a)).map(mapping).getOrElse(a)
+    })
+  }
+
+  /**
+   * Gather all the predicates and referenced attributes on different points of CTE references
+   * using pattern `ScanOperation` (which takes care of determinism) and combine those predicates
+   * and attributes that belong to the same CTE definition.
+   * For the same CTE definition, if any of its references does not have predicates, the combined
+   * predicate will be a TRUE literal, which means there will be no predicate push-down.
+   */
+  private def gatherPredicatesAndAttributes(plan: LogicalPlan, cteMap: CteMap): Unit = {
+    plan match {
+      case WithCTE(child, cteDefs) =>
+        cteDefs.zipWithIndex.foreach { case (cteDef, precedence) =>
+          gatherPredicatesAndAttributes(cteDef.child, cteMap)
+          cteMap.put(cteDef.id, (cteDef, precedence, Seq.empty, AttributeSet.empty))
+        }
+        gatherPredicatesAndAttributes(child, cteMap)
+
+      case ScanOperation(projects, predicates, ref: CTERelationRef) =>
+        val (cteDef, precedence, preds, attrs) = cteMap(ref.cteId)
+        val mapping = ref.output.zip(cteDef.output).map{ case (r, d) => r -> d }.toMap
+        val newPredicates = if (isTruePredicate(preds)) {
+          preds
+        } else {
+          val filteredPredicates = reverseMapExpressions(predicates.filter(_.find {
+            case s: SubqueryExpression => s.plan.find {
+              case r: CTERelationRef =>
+                !cteMap.contains(r.cteId) || cteMap(r.cteId)._2 >= precedence
```

Comment:
Can you add some comments to explain this condition?

Reply:
`!cteMap.contains(r.cteId)` is kind of a sanity check, as we should collect Def first before hitting its Ref. But I don't quite understand what `cteMap(r.cteId)._2 >= precedence` means.